contracts/adapters/CharmAdapter.sol

```diff
@@ -0,0 +1,388 @@
+        amountIn = market.buy(
+          optionType.isLongToken,
+          optionType.strikeIndex,
+          amount,
+          maxCost
+        );
+      }
+
+      emit Purchased(
+          msg.sender,
+          _name,
+          optionTerms.underlying,
+          amountIn,
+          0
+      );
+    }
+
+    /**
+     * @notice Exercises the options contract.
+     * @param options is the address of the options contract
+     * @param optionID is the ID of the option position in non fungible protocols like Hegic.
+     * @param amount is the amount of tokens or options contract to exercise. Only relevant for fungle protocols like Opyn
+     * @param recipient is the account that receives the exercised profits. This is needed since the adapter holds all the positions and the msg.sender is an instrument contract.
+     */
+    function exercise(
+        address options,
+        uint256 optionID,
+        uint256 amount,
+        address recipient
+    ) public payable override {
+
+      // Get strike index, whether is long
+      OptionType memory optionType = addressToOptionType[options];
+
+      //get token
+      IOptionToken token = IOptionToken(options);
+      //get market
+      IOptionMarket market = IOptionMarket(token.market());
+
+      uint256 profit = exerciseProfit(options, optionID, amount);
+      require(profit > 0, "Not profitable to exercise");
+
+      // if we are exercising but market has not settled, do it
+      if(market.isExpired() && !market.isSettled()){
+        market.settle();
+      }
+
+      uint256 amountOut = market.sell(optionType.isLongToken, optionType.strikeIndex, amount, profit);
+
+      //transfer over
+      market.baseToken().uniTransfer(payable(recipient), amountOut);
+
+      emit Exercised(
+          msg.sender,
+          options,
+          optionID,
+          amount,
+          amountOut
+      );
+    }
+
+    function createShort(ProtocolAdapterTypes.OptionTerms memory, uint256)
+        external
+        pure
+        override
+        returns (uint256)
+    {
+        return 0;
+    }
+
+    function closeShort() external pure override returns (uint256) {
+        return 0;
+    }
+
+    /**
+     * @notice Function to lookup oToken addresses.
+     * @param optionTerms is the terms of the option contract
+     */
+    function lookupOToken(ProtocolAdapterTypes.OptionTerms memory optionTerms)
+        public
+        view
+        returns (address token)
+    {
+        bool isPut =
+            optionTerms.optionType == ProtocolAdapterTypes.OptionType.Put;
+
+        //market.baseToken() is underlying asset if call. Strike currency if put. Represents ETH if equal to 0x0
+        address underlying = isPut == true ? optionTerms.strikeAsset : optionTerms.underlying;
+
+        //there should only be a collateral asset for charm if we are writing an option
+        bool isShort = optionTerms.collateralAsset == address(1) ? false : true;
```

Comment:
there should only be a collateral asset for charm if we are writing/selling an option, right? if so address(1) will designate buying instead of selling and anything else is buying

Reply:
Damn, this differentiation between long and short tokens is seriously breaking our OptionTerms abstraction 😿 

>there should only be a collateral asset for charm if we are writing/selling an option, right? if so address(1) will designate buying instead of selling and anything else is buying

I'm good with this approach 👍